oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/index/ElasticsearchIndexWriter.java

```diff
@@ -104,9 +107,36 @@ public void deleteDocuments(String path) throws IOException {
 
     @Override
     public boolean close(long timestamp) throws IOException {
-        // TODO : track index updates and return accordingly
-        // TODO : if/when we do async push, this is where to wait for those ops to complete
-        return false;
+        LOG.trace("Calling close on bulk processor {}", bulkProcessor);
+        bulkProcessor.close();
+        LOG.trace("Bulk Processor {} closed", bulkProcessor);
+
+        // bulkProcessor.close() calls the OakBulkProcessorListener.beforeBulk in a blocking manner
+        // indexUpdated would be unset there if it was false till now (not even a single update succeeded)
+        // in this case wait for sometime for the last OakBulkProcessorListener.afterBulk to be called
+        // where indexUpdated can possibly be set to true, return false in case of timeout.
+        // We don't wait in case indexUpdated is already set (This would be if any of the previous flushes for this processor
+        // were successfull i.e index was updated at least once)
```

Comment:
(optional) The logic seems correct to me. But since I see wait for operations to be completed and timeouts, I was wondering if some `java.util.concurrent` utility could be used to avoid race conditions (eg: BlockingQueue or maybe CountDownLatch).

Reply:
@fabriziofortino  - the indexUpdated Optional (that drives the logic) isn't shared across the bulkProcessor instances, so I was wondering what possible race condition could there be ?

Maybe the afterBulk callbacks for different bulk requests can cause problems ? One thing I can think of is - let's say the bulkRequest 1 is failing for all it's entries and it's in a retry loop and meanwhile bulkRequest 2 comes in and succeeds, and sets the indexUpdated Optional to true.
bulkRequest 1 in this case would then try to set it to false whenever it returns to the afterBulk callback - but the current logic won't allow it to do so (because if this flag is set to true once then it can't be set to false - because we just need 1 good index update to say that some write happened).

Or Maybe you had some other race condition in mind ?